<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Стоимость-автомобиля-на-вторичном-рынке" data-toc-modified-id="Стоимость-автомобиля-на-вторичном-рынке-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Стоимость автомобиля на вторичном рынке</a></span><ul class="toc-item"><li><span><a href="#Описание-работы" data-toc-modified-id="Описание-работы-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание работы</a></span><ul class="toc-item"><li><span><a href="#Постановка-задачи" data-toc-modified-id="Постановка-задачи-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>Постановка задачи</a></span></li><li><span><a href="#Описание-входных-данных" data-toc-modified-id="Описание-входных-данных-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Описание входных данных</a></span></li><li><span><a href="#Цели-и-этапы-решения-задачи" data-toc-modified-id="Цели-и-этапы-решения-задачи-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Цели и этапы решения задачи</a></span></li></ul></li></ul></li><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Изучение данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Просмотр-данных" data-toc-modified-id="Просмотр-данных-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Просмотр данных</a></span></li></ul></li><li><span><a href="#Обработка-данных" data-toc-modified-id="Обработка-данных-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка данных</a></span><ul class="toc-item"><li><span><a href="#Признак-year" data-toc-modified-id="Признак-year-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Признак year</a></span></li><li><span><a href="#Признак-make" data-toc-modified-id="Признак-make-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Признак make</a></span></li><li><span><a href="#Признак-model" data-toc-modified-id="Признак-model-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Признак model</a></span></li><li><span><a href="#Признак-trim" data-toc-modified-id="Признак-trim-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span>Признак trim</a></span></li><li><span><a href="#Признак-body" data-toc-modified-id="Признак-body-2.3.5"><span class="toc-item-num">2.3.5&nbsp;&nbsp;</span>Признак body</a></span></li><li><span><a href="#Признак-transmission" data-toc-modified-id="Признак-transmission-2.3.6"><span class="toc-item-num">2.3.6&nbsp;&nbsp;</span>Признак transmission</a></span></li><li><span><a href="#Признак-state" data-toc-modified-id="Признак-state-2.3.7"><span class="toc-item-num">2.3.7&nbsp;&nbsp;</span>Признак state</a></span></li><li><span><a href="#Признак-condition" data-toc-modified-id="Признак-condition-2.3.8"><span class="toc-item-num">2.3.8&nbsp;&nbsp;</span>Признак condition</a></span></li><li><span><a href="#Признак-odometer" data-toc-modified-id="Признак-odometer-2.3.9"><span class="toc-item-num">2.3.9&nbsp;&nbsp;</span>Признак odometer</a></span></li><li><span><a href="#Признак-color-и-interior" data-toc-modified-id="Признак-color-и-interior-2.3.10"><span class="toc-item-num">2.3.10&nbsp;&nbsp;</span>Признак color и interior</a></span></li><li><span><a href="#Признак-sellingprice" data-toc-modified-id="Признак-sellingprice-2.3.11"><span class="toc-item-num">2.3.11&nbsp;&nbsp;</span>Признак sellingprice</a></span></li><li><span><a href="#Признак-saledate" data-toc-modified-id="Признак-saledate-2.3.12"><span class="toc-item-num">2.3.12&nbsp;&nbsp;</span>Признак saledate</a></span></li><li><span><a href="#Удалим-лишние-столбцы" data-toc-modified-id="Удалим-лишние-столбцы-2.3.13"><span class="toc-item-num">2.3.13&nbsp;&nbsp;</span>Удалим лишние столбцы</a></span></li></ul></li></ul></li><li><span><a href="#Моделирование" data-toc-modified-id="Моделирование-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Моделирование</a></span><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Колирование-категориальных-переменных" data-toc-modified-id="Колирование-категориальных-переменных-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Колирование категориальных переменных</a></span></li><li><span><a href="#Масштабирование-признаков" data-toc-modified-id="Масштабирование-признаков-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Масштабирование признаков</a></span></li></ul></li><li><span><a href="#DecisionTreeRegressor" data-toc-modified-id="DecisionTreeRegressor-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>DecisionTreeRegressor</a></span></li><li><span><a href="#HistGradientBoostingRegressor" data-toc-modified-id="HistGradientBoostingRegressor-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>HistGradientBoostingRegressor</a></span></li><li><span><a href="#CatBoostRegressor" data-toc-modified-id="CatBoostRegressor-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>CatBoostRegressor</a></span></li><li><span><a href="#RandomForestRegressor" data-toc-modified-id="RandomForestRegressor-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>RandomForestRegressor</a></span></li></ul></li><li><span><a href="#Рассчет-выбранной-модели-на-тестовых-данных" data-toc-modified-id="Рассчет-выбранной-модели-на-тестовых-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Рассчет выбранной модели на тестовых данных</a></span><ul class="toc-item"><li><span><a href="#Кодирование-и-масштабирование-данных-для-полного-датасета" data-toc-modified-id="Кодирование-и-масштабирование-данных-для-полного-датасета-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Кодирование и масштабирование данных для полного датасета</a></span></li><li><span><a href="#Расчет-модели-с-выбранными-гиперпараметрами" data-toc-modified-id="Расчет-модели-с-выбранными-гиперпараметрами-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Расчет модели с выбранными гиперпараметрами</a></span></li><li><span><a href="#Формирование-и-сохранение-итогового-Датасета-sample_submission" data-toc-modified-id="Формирование-и-сохранение-итогового-Датасета-sample_submission-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Формирование и сохранение итогового Датасета sample_submission</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Стоимость автомобиля на вторичном рынке

## Описание работы

### Постановка задачи

***Заказчик***: Агентство по подбору автомобилей со вторичного рынка.

Необходимо разработать модель, которая поможет спрогнозировать стоимость автомобиля на вторичном рынке. Тем самым заказчик планирует создавать перечини автомобилей по пожеланиям клиентов учитывая как ограничения по стоимости, так пожелания по маркам.

Агентство объявило конкурс на платформе Kaggle. Для тестирования модели необходимо отправить тестовую выборку на платформу для получения решающего результата.

В качестве данных для обучения модели получены ретроспективные данные о продажах автомобилей со вторичного рынка за 2014-2015 года. Также получены данные для тестирования модели.

Для решения задачи использовать язык Python (Jupyter Notebook).


### Описание входных данных

**Дано:**
1. train.csv - информация о продажах (~440000) автомобилей с аукционов, которые будут использоваться в качестве обучающих данных.
2. test.csv - информация о продажах (~110000) автомобилей с аукционов, которые будут использоваться в качестве тестовых данных. данные не имеют поля 'sellingprice', при загрузке результата на платформу будет получен результат

sellingprice - Целевой признак. Для каждого автомобиля предскажите числовое значение стоимости автомобиля.

**Описание полей данных:**
1. 'year' - год производства
2. 'make' - производитель
3. 'model' - модель
4. 'trim' - модификация
5. 'body' - тип кузова
6. 'transmission' - тип КПП
7. 'vin' - идентификатор (вин)
8. 'state' - штат регистрации
9. 'condition' - состояние по шкале (1-5)
10. 'odometer' - пробег в милях
11. 'color' - цвет кузова
12. 'interior' - цвет интерьера
13. 'seller' - продавец
14. 'sellingprice' - стоимость продажи
15. 'saledate' - дата продажи

### Цели и этапы решения задачи

**Цель:** реализовать модель позволяющую выяснить предварительную оценку автомобиля.

**Этапы решения задачи:**
1.	Изучить данные
2.	Обработать данные: заполнить пропуски, вычленить выбросы, убрать явняе и неявные дубликаты.
3.	Выделить из обучающей выборки часть данных для тестирования модели
4.	Обучить несколько моделей на полученных данных: DecisionTreeRegressor, RandomForestRegressor, GradientBoostingRegressor, CatBoostRegressor
5.	Выбрать наилучшие модели и протестировать их на выделенной части данных. Для отбора модели использовать GridSearchCV, главная метрика mean_absolute_percentage_error.
6.	Обучить выбранную модель на всех данных обучающей выборки
7.	Применить модель для тестовой выборки от заказчика
8.	Выложить результат на платформу Kaggle в качестве csv файла в формате vin - price
9.	Написать отчет о проделанной работе.

# Изучение данных

## Импорт библиотек

In [142]:
!pip install scikit-learn==1.1.3
!pip install catboost
!pip show scikit-learn

Name: scikit-learn
Version: 1.1.3
Summary: A set of python modules for machine learning and data mining
Home-page: http://scikit-learn.org
Author: 
Author-email: 
License: new BSD
Location: c:\users\fenya\anaconda3\lib\site-packages
Requires: joblib, threadpoolctl, numpy, scipy
Required-by: scikit-learn-intelex


In [143]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import (RandomForestRegressor, GradientBoostingRegressor, HistGradientBoostingRegressor)
from catboost import CatBoostRegressor
from sklearn.metrics import mean_absolute_percentage_error
from sklearn.preprocessing import (OrdinalEncoder, StandardScaler)
from sklearn.model_selection import (train_test_split, GridSearchCV)

## Загрузка данных

In [144]:
train = pd.read_csv(r'train.csv')
test_df = pd.read_csv(r'test.csv')
test = test_df.copy()

### Просмотр данных

In [145]:
display(train.head(3))
train.info()

year    make   model    trim   body transmission                vin state  \
0  2011    Ford    Edge     SEL    suv    automatic  2fmdk3jc4bba41556    md   
1  2014    Ford  Fusion      SE  Sedan    automatic  3fa6p0h75er208976    mo   
2  2012  Nissan  Sentra  2.0 SL  sedan    automatic  3n1ab6ap4cl698412    nj   

   condition  odometer  color interior                 seller  sellingprice  \
0        4.2  111041.0  black    black     santander consumer         12500   
1        3.5   31034.0  black    black  ars/avis budget group         14500   
2        2.2   35619.0  black    black     nissan-infiniti lt          9100   

                                  saledate  
0  Tue Jun 02 2015 02:30:00 GMT-0700 (PDT)  
1  Wed Feb 25 2015 02:00:00 GMT-0800 (PST)  
2  Wed Jun 10 2015 02:30:00 GMT-0700 (PDT)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 440236 entries, 0 to 440235
Data columns (total 15 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   year          440236 non-null  int64  
 1   make          432193 non-null  object 
 2   model         432113 non-null  object 
 3   trim          431899 non-null  object 
 4   body          429843 non-null  object 
 5   transmission  388775 non-null  object 
 6   vin           440236 non-null  object 
 7   state         440236 non-null  object 
 8   condition     430831 non-null  float64
 9   odometer      440167 non-null  float64
 10  color         439650 non-null  object 
 11  interior      439650 non-null  object 
 12  seller        440236 non-null  object 
 13  sellingprice  440236 non-null  int64  
 14  saledate      440236 non-null  object 
dtypes: float64(2), int64(2), object(11)
memory usage: 50.4+ MB


Данные содержит 440236 объектов, каждый из объектов имеет 15 признаков.

Из постановки задачи видно, что столбец sellingprice будет эндогенной (зависимой) переменной модели или откликом, а остальные 14 признаков будут экзогенным(независимыми) переменными.

Данные содержат пропуски и неявные дубликаты.

## Обработка данных

### Признак year

In [146]:
print(f'Kоличество пропусков: ',train['year'].isna().sum())
print(f'Kоличество категорий: ',train['year'].nunique())
print(f'Список категорий: ',train.sort_values(by = 'year')['year'].unique())

Kоличество пропусков:  0
Kоличество категорий:  34
Список категорий:  [1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993 1994 1995
 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007 2008 2009
 2010 2011 2012 2013 2014 2015]


Датасет содержит данные с автомобилями с 1982 по  2015 года выпуска. Признак категориальный, пропусков нет.

### Признак make

In [147]:
print(f'Kоличество пропусков: ',train['make'].isna().sum())
print(f'Kоличество категорий: ',train['make'].nunique())
print(f'Список категорий: ',train.sort_values(by = 'make')['make'].unique())

Kоличество пропусков:  8043
Kоличество категорий:  92
Список категорий:  ['Acura' 'Aston Martin' 'Audi' 'BMW' 'Bentley' 'Buick' 'Cadillac'
 'Chevrolet' 'Chrysler' 'Daewoo' 'Dodge' 'FIAT' 'Ferrari' 'Fisker' 'Ford'
 'GMC' 'Geo' 'HUMMER' 'Honda' 'Hyundai' 'Infiniti' 'Isuzu' 'Jaguar' 'Jeep'
 'Kia' 'Lamborghini' 'Land Rover' 'Lexus' 'Lincoln' 'MINI' 'Maserati'
 'Mazda' 'Mercedes-Benz' 'Mercury' 'Mitsubishi' 'Nissan' 'Oldsmobile'
 'Plymouth' 'Pontiac' 'Porsche' 'Ram' 'Rolls-Royce' 'Saab' 'Saturn'
 'Scion' 'Subaru' 'Suzuki' 'Tesla' 'Toyota' 'Volkswagen' 'Volvo' 'acura'
 'airstream' 'audi' 'bmw' 'buick' 'cadillac' 'chevrolet' 'chrysler'
 'dodge' 'dodge tk' 'dot' 'ford' 'ford truck' 'gmc' 'gmc truck' 'honda'
 'hyundai' 'jeep' 'kia' 'land rover' 'landrover' 'lexus' 'lincoln'
 'maserati' 'mazda' 'mazda tk' 'mercedes' 'mercedes-b' 'mercury'
 'mitsubishi' 'nissan' 'oldsmobile' 'plymouth' 'pontiac' 'porsche' 'smart'
 'subaru' 'suzuki' 'toyota' 'volkswagen' 'vw' nan]


Признак содержит неявные дубликаты и пропуски. Пропуски заполним значением 'unknown'. Удалим неявные дубликаты. Тестовую выборку обработаем по такой же схеме.

In [148]:
#remove capital letters
train['make'] = train['make'].str.lower()
test['make'] = test['make'].str.lower()

train['make'] = train['make'].fillna('unknown')
test['make'] = test['make'].fillna('unknown')

wrong_words = ['-benz', 'vw', '-b', ' tk', 'chev ', ' truck']
good_name = ['', 'volkswagen', '', ' truck', 'chevrolet ', '']
for i in range(len(wrong_words)):
    train['make'] = train['make'].str.replace(wrong_words[i],good_name[i])
    test['make'] = test['make'].str.replace(wrong_words[i],good_name[i])


In [149]:
print(f'Kоличество пропусков: ',train['make'].isna().sum())
print(f'Kоличество категорий: ',train['make'].nunique())
print(f'Список категорий: ',train.sort_values(by = 'make')['make'].unique())

Kоличество пропусков:  0
Kоличество категорий:  56
Список категорий:  ['acura' 'airstream' 'aston martin' 'audi' 'bentley' 'bmw' 'buick'
 'cadillac' 'chevrolet' 'chrysler' 'daewoo' 'dodge' 'dot' 'ferrari' 'fiat'
 'fisker' 'ford' 'geo' 'gmc' 'honda' 'hummer' 'hyundai' 'infiniti' 'isuzu'
 'jaguar' 'jeep' 'kia' 'lamborghini' 'land rover' 'landrover' 'lexus'
 'lincoln' 'maserati' 'mazda' 'mercedes' 'mercury' 'mini' 'mitsubishi'
 'nissan' 'oldsmobile' 'plymouth' 'pontiac' 'porsche' 'ram' 'rolls-royce'
 'saab' 'saturn' 'scion' 'smart' 'subaru' 'suzuki' 'tesla' 'toyota'
 'unknown' 'volkswagen' 'volvo']


После обработки датасет содержит информацию о автомобилях 55 марок. ПРопуски заполнены значением unknown

### Признак model

In [150]:
print(f'Kоличество пропусков: ',train['model'].isna().sum())
print(f'Kоличество категорий: ',train['model'].nunique())
# print(f'Список категорий: ',train.sort_values(by = 'model')['model'].unique())

Kоличество пропусков:  8123
Kоличество категорий:  959


Признак содержит неявные дубликаты и пропуски. Пропуски заполним значением 'unknown'. Удалим неявные дубликаты. Тестовую выборку обработаем по такой же схеме.

In [151]:
train['model'] = train['model'].str.lower()
test['model'] = test['model'].str.lower()

train['model'] = train['model'].fillna('unknown')
test['model'] = test['model'].fillna('unknown')

wrong_words = ['-', ' ']
good_name = ['', '']
for i in range(len(wrong_words)):
    train['model'] = train['model'].str.replace(wrong_words[i],good_name[i])
    test['model'] = test['model'].str.replace(wrong_words[i],good_name[i])

In [152]:
print(f'Kоличество пропусков: ',train['model'].isna().sum())
print(f'Kоличество категорий: ',train['model'].nunique())
# print(f'Список категорий: ',train.sort_values(by = 'model')['model'].unique())

Kоличество пропусков:  0
Kоличество категорий:  830


После обработки датасет содержит информацию о 829 моделях автомобилей. Пропуски заполнены значением unknown

### Признак trim

In [153]:
print(f'Kоличество пропусков: ',train['trim'].isna().sum())
print(f'Kоличество категорий: ',train['trim'].nunique())
# print(f'Список категорий: ',train.sort_values(by = 'trim')['trim'].unique())

Kоличество пропусков:  8337
Kоличество категорий:  1922


Признак содержит неявные дубликаты и пропуски. Пропуски заполним значением 'unknown'. Удалим неявные дубликаты. Тестовую выборку обработаем по такой же схеме.

In [154]:
train['trim'] = train['trim'].str.lower()
test['trim'] = test['trim'].str.lower()

train['trim'] = train['trim'].fillna('unknown')
test['trim'] = test['trim'].fillna('unknown')

wrong_words = ['-', ' ']
good_name = ['','']
for i in range(len(wrong_words)):
    train['trim'] = train['trim'].str.replace(wrong_words[i],good_name[i])
    test['trim'] = test['trim'].str.replace(wrong_words[i],good_name[i])

In [155]:
print(f'Kоличество пропусков: ',train['trim'].isna().sum())
print(f'Kоличество категорий: ',train['trim'].nunique())
# print(f'Список категорий: ',train.sort_values(by = 'trim')['trim'].unique())

Kоличество пропусков:  0
Kоличество категорий:  1813


После обработки датасет содержит информацию о 1812 модификациях автомобилей. Пропуски заполнены значением unknown

### Признак body

In [156]:
print(f'Kоличество пропусков: ',train['body'].isna().sum())
print(f'Kоличество категорий: ',train['body'].nunique())
# print(f'Список категорий: ',train.sort_values(by = 'body')['body'].unique())

Kоличество пропусков:  10393
Kоличество категорий:  85


Признак содержит неявные дубликаты и пропуски. Пропуски заполним значением 'unknown'. Удалим неявные дубликаты. Тестовую выборку обработаем по такой же схеме.

In [157]:
train['body'] = train['body'].str.lower()
test['body'] = test['body'].str.lower()

train['body'] = train['body'].fillna('unknown')
test['body'] = test['body'].fillna('unknown')

wrong_words = ['-', ' ']
good_name = ['','']
for i in range(len(wrong_words)):
    train['body'] = train['body'].str.replace(wrong_words[i],good_name[i])
    test['body'] = test['body'].str.replace(wrong_words[i],good_name[i])

In [158]:
print(f'Kоличество пропусков: ',train['body'].isna().sum())
print(f'Kоличество категорий: ',train['body'].nunique())
# print(f'Список категорий: ',train.sort_values(by = 'body')['body'].unique())

Kоличество пропусков:  0
Kоличество категорий:  45


После обработки датасет содержит информацию о автомобилях 44 вариантов корпусов автомобилей. Пропуски заполнены значением unknown.

### Признак transmission

In [159]:
print(f'Kоличество пропусков: ',train['transmission'].isna().sum())
print(f'Kоличество категорий: ',train['transmission'].nunique())
print(f'Список категорий: ',train.sort_values(by = 'transmission')['transmission'].unique())

Kоличество пропусков:  51461
Kоличество категорий:  2
Список категорий:  ['automatic' 'manual' nan]


Признак содержит пропуски. Пропуски заполним значением 'unknown'. Тестовую выбрку обработаем по такой же схеме.

In [160]:
train['transmission'] = train['transmission'].fillna('unknown')
test['transmission'] = test['transmission'].fillna('unknown')

### Признак state

In [161]:
print(f'Kоличество пропусков: ',train['state'].isna().sum())
print(f'Kоличество категорий: ',train['state'].nunique())
print(f'Список категорий: ',train.sort_values(by = 'state')['state'].unique())

Kоличество пропусков:  0
Kоличество категорий:  38
Список категорий:  ['ab' 'al' 'az' 'ca' 'co' 'fl' 'ga' 'hi' 'il' 'in' 'la' 'ma' 'md' 'mi'
 'mn' 'mo' 'ms' 'nc' 'ne' 'nj' 'nm' 'ns' 'nv' 'ny' 'oh' 'ok' 'on' 'or'
 'pa' 'pr' 'qc' 'sc' 'tn' 'tx' 'ut' 'va' 'wa' 'wi']


Признак не содержит дубликатов и пропусков

### Признак condition

In [162]:
print(f'Kоличество пропусков: ',train['condition'].isna().sum())

Kоличество пропусков:  9405


Признак непрерывный содержит значение состояние автомобиля по шкале от 1 до 5. Проверим экстремальные значения и заполним пропуски исходя из медианы по году выпуска. Тестовую выборку обработает по такому же сценарию.

In [163]:
for year in train.sort_values(by = 'year')['year'].unique():
    train['condition'] = train['condition'].fillna(train.loc[train['year'] == year]
                                                       ['condition'].median())

for year in test.sort_values(by = 'year')['year'].unique():
    test['condition'] = test['condition'].fillna(test.loc[test['year'] == year]
                                                       ['condition'].median())

In [164]:
print('Максимум: ', train['condition'].max())
print('Минимум: ', train['condition'].min())

Максимум:  5.0
Минимум:  1.0


### Признак odometer

In [165]:
print(f'Kоличество пропусков: ',train['odometer'].isna().sum())

Kоличество пропусков:  69


Признак непрерывный содержит значение пробега автомобиля. Проверим минимальный пробег (он должен быть положительным и хотябы больше 50 миль). Тестовую выборку обработает по такому же сценарию. Заполним пропуски исходя из медианы по году выпуска. Тестовую выборку обработает по такому же сценарию.

In [166]:
train = train.query('~(odometer<10)')

for year in train.sort_values(by = 'year')['year'].unique():
    train['odometer'] = train['odometer'].fillna(train.loc[train['year'] == year]
                                                       ['odometer'].median())

for year in test.sort_values(by = 'year')['year'].unique():
    test['odometer'] = test['odometer'].fillna(test.loc[test['year'] == year]
                                                       ['odometer'].median())

### Признак color и interior

In [167]:
print(f'Kоличество пропусков color: ',train['color'].isna().sum())
print(f'Kоличество категорий color: ',train['color'].nunique())
print(f'Список категорий color: ',train.sort_values(by = 'color')['color'].unique())

Kоличество пропусков color:  574
Kоличество категорий color:  20
Список категорий color:  ['beige' 'black' 'blue' 'brown' 'burgundy' 'charcoal' 'gold' 'gray'
 'green' 'lime' 'off-white' 'orange' 'pink' 'purple' 'red' 'silver'
 'turquoise' 'white' 'yellow' '—' nan]


In [168]:
print(f'Kоличество пропусков interior: ',train['interior'].isna().sum())
print(f'Kоличество категорий interior: ',train['interior'].nunique())
print(f'Список категорий interior: ',train.sort_values(by = 'interior')['interior'].unique())

Kоличество пропусков interior:  574
Kоличество категорий interior:  17
Список категорий interior:  ['beige' 'black' 'blue' 'brown' 'burgundy' 'gold' 'gray' 'green'
 'off-white' 'orange' 'purple' 'red' 'silver' 'tan' 'white' 'yellow' '—'
 nan]


Признаки категориальные и содержат пропуски. Заполним пропуски и - значением unknown.

In [169]:
train['color'] = train['color'].fillna('unknown')
test['color'] = test['color'].fillna('unknown')
train['interior'] = train['interior'].fillna('unknown')
test['interior'] = test['interior'].fillna('unknown')

train['color'] = train['color'].replace('—','unknown')
test['color'] = test['color'].replace('—','unknown')
train['interior'] = train['interior'].replace('—','unknown')
test['interior'] = test['interior'].replace('—','unknown')

In [170]:
print(f'Kоличество пропусков color: ',train['color'].isna().sum())
print(f'Kоличество категорий color: ',train['color'].nunique())
print(f'Список категорий color: ',train.sort_values(by = 'color')['color'].unique())

Kоличество пропусков color:  0
Kоличество категорий color:  20
Список категорий color:  ['beige' 'black' 'blue' 'brown' 'burgundy' 'charcoal' 'gold' 'gray'
 'green' 'lime' 'off-white' 'orange' 'pink' 'purple' 'red' 'silver'
 'turquoise' 'unknown' 'white' 'yellow']


In [171]:
print(f'Kоличество пропусков interior: ',train['interior'].isna().sum())
print(f'Kоличество категорий interior: ',train['interior'].nunique())
print(f'Список категорий interior: ',train.sort_values(by = 'interior')['interior'].unique())

Kоличество пропусков interior:  0
Kоличество категорий interior:  17
Список категорий interior:  ['beige' 'black' 'blue' 'brown' 'burgundy' 'gold' 'gray' 'green'
 'off-white' 'orange' 'purple' 'red' 'silver' 'tan' 'unknown' 'white'
 'yellow']


### Признак sellingprice

Исключим из данных машины с минимальными и максимальными ценами

In [172]:
train = train.query('~(sellingprice<250)')
train = train.query('~(sellingprice>185000)')

### Признак saledate

При помощи этого признака определим возраст автомобиля в момент продажи

In [173]:
train['saledate'] = [str(i).split(' GMT')[0] for i in train['saledate'].values]
train['saledate'] = pd.to_datetime(train['saledate'], format='%a %b %d %Y %H:%M:%S')
test['saledate'] = [str(i).split(' GMT')[0] for i in test['saledate'].values]
test['saledate'] = pd.to_datetime(test['saledate'], format='%a %b %d %Y %H:%M:%S')

train['saleyear'] = train['saledate'].dt.year
train['age'] = train['saleyear'] - train['year'] + 1
test['saleyear'] = test['saledate'].dt.year
test['age'] = test['saleyear'] - test['year'] + 1

Удалим столбцы vin - так как это идентификационный номер, year, saledate и saleyear - так как информация из этих параметров уже учтена в переменной age, а также seller, так как даже после обработки в переменной больше 10 тысяч категорий.

### Удалим лишние столбцы

In [174]:
train = train.drop(['year','vin', 'seller', 'saledate', 'saleyear'], axis=1)
test = test.drop(['year', 'vin', 'seller', 'saledate', 'saleyear'], axis=1)
train.head(1)
test.head(1)

make model  trim   body transmission state  condition  odometer  \
0  cadillac   cts  base  sedan    automatic    ca        2.7  116970.0   

    color interior  age  
0  silver    black   11

# Моделирование

## Подготовка данных

Для корректного отбора моделей отделим от выборки train 40% наблюдений и будем тестировать результат моделирования на этой псевдотестовой выборке. Назовём её valid.

In [175]:
X = train.drop(['sellingprice'], axis=1)
y = train['sellingprice']

X_train, X_valid, y_train, y_valid = train_test_split(
    X, y, test_size=0.4, random_state=12345)

X_train

make         model              trim         body transmission  \
197672      lexus         es350              base        sedan    automatic   
60826      nissan          leaf                sl    hatchback    automatic   
316968        kia        optima                lx        sedan      unknown   
8776    chevrolet        impala           lsfleet        sedan    automatic   
310695       ford        fusion                se        sedan    automatic   
...           ...           ...               ...          ...          ...   
348659        bmw       3series              328i  convertible    automatic   
396666       ford  eserieswagon  e350superdutyxlt          van    automatic   
393740      dodge       journey               sxt          suv    automatic   
77510        ford        taurus               ses        sedan    automatic   
218247      dodge  grandcaravan               sxt      minivan      unknown   

       state  condition  odometer   color interior  age  
197672    az        3.9   36812.0    gray     gray    2  
60826     ca        4.4   19723.0   white      tan    4  
316968    pa        4.6   26206.0   black     gray    4  
8776      ma        3.5   60457.0   white     gray    3  
310695    ne        3.6   40132.0   white    black    2  
...      ...        ...       ...     ...      ...  ...  
348659    fl        3.6   59160.0    gray    beige    8  
396666    fl        4.1   17543.0   white     gray    2  
393740    fl        4.0   19964.0  silver    black    2  
77510     fl        1.9  104690.0   black     gray   14  
218247    pa        4.2   67100.0   white    black    3  

[263298 rows x 11 columns]

### Колирование категориальных переменных

Обработаем категориальные признаки прямым колированием с помощью метода OrdinalEncoder.

In [176]:
X_train_ore = X_train.copy()
X_valid_ore = X_valid.copy()

categor = ['make', 'model', 'trim', 'body', 'transmission','state','color','interior']
encoder_ore = OrdinalEncoder(handle_unknown = 'use_encoded_value', unknown_value = -1)
encoder_ore.fit(X_train_ore[categor])

X_train_ore[encoder_ore.get_feature_names_out()] = encoder_ore.transform(X_train_ore[categor])
X_valid_ore[encoder_ore.get_feature_names_out()] = encoder_ore.transform(X_valid_ore[categor])

### Масштабирование признаков

Неприрывные признаки проскалируем методом StandardScaler

In [177]:
numeric = ['condition', 'odometer', 'age']
scaler = StandardScaler()
scaler.fit(X_train_ore[numeric])

X_train_ore[numeric] = scaler.transform(X_train_ore[numeric])
X_valid_ore[numeric] = scaler.transform(X_valid_ore[numeric])

## DecisionTreeRegressor

In [178]:
model = DecisionTreeRegressor(random_state = 123456)

parameters = {'max_depth': range(10,100,10), 
              'min_samples_leaf': [1,3,5,7,9], 
              'max_features': [2,4,6,10]
             }

modelgrid = GridSearchCV(model, 
                         parameters, 
                         scoring = 'neg_mean_absolute_percentage_error', 
                         cv = 5,
                         n_jobs = 6)

modelgrid = modelgrid.fit(X_train_ore, y_train)
# modelgrid = modelgrid.fit(X, y)

print(modelgrid.best_params_)

pred_train_ore = modelgrid.predict(X_train_ore)
mae_train_ore = round(mean_absolute_percentage_error(y_train, pred_train_ore),3)
print('mae_train_ore = ', mae_train_ore)

pred_valid_ore = modelgrid.predict(X_valid_ore)
mae_valid_ore = round(mean_absolute_percentage_error(y_valid, pred_valid_ore),3)
print('mae_valid_ore = ', mae_valid_ore)

{'max_depth': 40, 'max_features': 10, 'min_samples_leaf': 7}
mae_train_ore =  0.149
mae_valid_ore =  0.221


Модель отбирается с помощью метрики neg_mean_absolute_percentage_error (MAPE). 

Модель DecisionTreeRegressor, которая дала лучший результат имеет настройки: 'max_depth': 40, 'max_features': 10, 'min_samples_leaf': 7
На валидационноы выборке модель показывает значение метрики равное 0.221

## HistGradientBoostingRegressor

In [179]:
model = HistGradientBoostingRegressor(random_state = 123456)

parameters = {'max_depth': range(10,100,10), 
              'min_samples_leaf': [10,20,50,100], 
              'max_iter': [70,100,150,500]
             }

modelgrid = GridSearchCV(model, 
                         parameters, 
                         scoring = 'neg_mean_absolute_percentage_error', 
                         cv = 5,
                         n_jobs = 6)

modelgrid = modelgrid.fit(X_train_ore, y_train)
# modelgrid = modelgrid.fit(X, y)

print(modelgrid.best_params_)

pred_train_ore = modelgrid.predict(X_train_ore)
mae_train_ore = round(mean_absolute_percentage_error(y_train, pred_train_ore),3)
print('mae_train_ore = ', mae_train_ore)

pred_valid_ore = modelgrid.predict(X_valid_ore)
mae_valid_ore = round(mean_absolute_percentage_error(y_valid, pred_valid_ore),3)
print('mae_valid_ore = ', mae_valid_ore)

{'max_depth': 20, 'max_iter': 500, 'min_samples_leaf': 10}
mae_train_ore =  0.188
mae_valid_ore =  0.194


Модель HistGradientBoostingRegressor, которая дала лучший результат имеет настройки: 'max_depth': 20, 'max_iter': 500, 'min_samples_leaf': 10
На валидационной выборке модель показывает значение метрики равное 0.194

## CatBoostRegressor

In [180]:
model = CatBoostRegressor(random_state = 123456, metric_period=100)

parameters = {'max_depth'     : [12,14],
              'learning_rate' : [0.3,0.4],
              'iterations'    : [500,700]
              }

modelgrid = GridSearchCV(model, 
                         parameters, 
                         scoring = 'neg_mean_absolute_percentage_error', 
                         cv = 5,
                         n_jobs = 6)

modelgrid = modelgrid.fit(X_train_ore, y_train)
# modelgrid = modelgrid.fit(X, y)

print(modelgrid.best_params_)

pred_train_ore = modelgrid.predict(X_train_ore)
mae_train_ore = round(mean_absolute_percentage_error(y_train, pred_train_ore),3)
print('mae_train_ore = ', mae_train_ore)

pred_valid_ore = modelgrid.predict(X_valid_ore)
mae_valid_ore = round(mean_absolute_percentage_error(y_valid, pred_valid_ore),3)
print('mae_valid_ore = ', mae_valid_ore)

C:\Users\Fenya\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
2 fits failed out of a total of 40.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
2 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Fenya\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Fenya\anaconda3\lib\site-packages\catboost\core.py", line 5734, in fit
    return self._fit(X, y, cat_features, text_features, embedding_features, None, sample_weight, None, None, None, None, baseline,
  File "C:\Users\Fenya\anaconda3\lib\site-packages\catboost\core.py", line 2357, in _fit
    

0:	learn: 8044.1345083	total: 200ms	remaining: 2m 19s
100:	learn: 2210.7852617	total: 6.22s	remaining: 36.9s
200:	learn: 1860.4705623	total: 12.3s	remaining: 30.5s
300:	learn: 1679.4556762	total: 18.3s	remaining: 24.3s
400:	learn: 1548.5859494	total: 24.8s	remaining: 18.5s
500:	learn: 1446.8111809	total: 30.9s	remaining: 12.3s
600:	learn: 1364.1895852	total: 36.8s	remaining: 6.07s
699:	learn: 1296.7664697	total: 42.6s	remaining: 0us
{'iterations': 700, 'learning_rate': 0.3, 'max_depth': 12}
mae_train_ore =  0.122
mae_valid_ore =  0.168


Модель CatBoostRegressor, которая дала лучший результат имеет настройки: 'iterations': 700, 'learning_rate': 0.3, 'max_depth': 12. На валидационной выборке модель показывает значение метрики равное 0.168

## RandomForestRegressor

In [181]:
model = RandomForestRegressor(random_state = 123456)

parameters = {'n_estimators' : [70,100, 150],
              'max_depth' : [10, 15]
              }

modelgrid = GridSearchCV(model, 
                         parameters, 
                         scoring = 'neg_mean_absolute_percentage_error', 
                         cv = 5,
                         n_jobs = 6)

modelgrid = modelgrid.fit(X_train_ore, y_train)
# modelgrid = modelgrid.fit(X, y)

print(modelgrid.best_params_)

pred_train_ore = modelgrid.predict(X_train_ore)
mae_train_ore = round(mean_absolute_percentage_error(y_train, pred_train_ore),3)
print('mae_train_ore = ', mae_train_ore)

pred_valid_ore = modelgrid.predict(X_valid_ore)
mae_valid_ore = round(mean_absolute_percentage_error(y_valid, pred_valid_ore),3)
print('mae_valid_ore = ', mae_valid_ore)

{'max_depth': 15, 'n_estimators': 150}
mae_train_ore =  0.187
mae_valid_ore =  0.229


Модель RandomForestRegressor, которая дала лучший результат имеет настройки: 'max_depth': 15, 'n_estimators': 150. На валидационной выборке модель показывает значение метрики равное 0.229.

Лучшей моделью стала CatBoostRegressor с гиперпараметрами 'iterations': 700, 'learning_rate': 0.3, 'max_depth': 12. Используем её для дальнейшего прогнозирования.

# Рассчет выбранной модели на тестовых данных

## Кодирование и масштабирование данных для полного датасета

In [182]:
X_ore = X.copy()
test_ore = test.copy()

categor = ['make', 'model', 'trim', 'body', 'transmission','state','color','interior']
encoder_ore = OrdinalEncoder(handle_unknown = 'use_encoded_value', unknown_value = -1)
encoder_ore.fit(X_ore[categor])

X_ore[encoder_ore.get_feature_names_out()] = encoder_ore.transform(X_ore[categor])
test_ore[encoder_ore.get_feature_names_out()] = encoder_ore.transform(test_ore[categor])


numeric = ['condition', 'odometer', 'age']
scaler = StandardScaler()
scaler.fit(X_ore[numeric])

X_ore[numeric] = scaler.transform(X_ore[numeric])
test_ore[numeric] = scaler.transform(test_ore[numeric])

## Расчет модели с выбранными гиперпараметрами

In [183]:
best_model = CatBoostRegressor(random_state = 123456, metric_period=100, max_depth = 12, learning_rate = 0.3, iterations = 700)
best_model = best_model.fit(X_ore, y)

best_model_pred = best_model.predict(X_ore)
best_model_mae = round(mean_absolute_percentage_error(y, best_model_pred),3)
print('best_model_mae = ', best_model_mae)

0:	learn: 7982.0658808	total: 72.6ms	remaining: 50.7s
100:	learn: 2225.6362204	total: 7.06s	remaining: 41.9s
200:	learn: 1898.0229470	total: 14s	remaining: 34.8s
300:	learn: 1737.4371459	total: 21.1s	remaining: 28s
400:	learn: 1630.1595514	total: 28.3s	remaining: 21.1s
500:	learn: 1547.4421206	total: 35.3s	remaining: 14s
600:	learn: 1478.0840443	total: 42.4s	remaining: 6.99s
699:	learn: 1423.8951646	total: 49.4s	remaining: 0us
best_model_mae =  0.131


## Формирование и сохранение итогового Датасета sample_submission

In [184]:
sample_submission = pd.DataFrame()
sample_submission['vin'] = test_df['vin']
sample_submission['sellingprice'] = best_model.predict(test_ore)
sample_submission.head(3)

vin  sellingprice
0  1g6dp567450124779   4825.564855
1  1gtw7fca7e1902207  21522.772355
2  jn8az1mw6dw303497  19765.224536

In [188]:
sample_submission.to_csv('sample_submission.csv', index = False)

# Выводы

В результате проделанной работы были сделаны следующие шаги:
1. Загружены и обработаны данные
2. Обработанные данные были закодированы и отмасштабированы
3. Было обучено 4 вида регрессионных моделей: DecisionTreeRegressor, HistGradientBoostingRegressor, CatBoostRegressor и RandomForestRegressor
4. При помощи метода GridSearchCV с помощью метрики MAPE была выбрана модель CatBoostRegressor.
5. Результирующая модель была применена к тестовой выборке.
6. Результат был загружен на платформу kaggle, результат прогнозирования 18.35694
